File: mycuration/src/test/java/com/phicdy/mycuration/presentation/presenter/RegisterFilterPresenterTest.java
package com.phicdy.mycuration.presentation.presenter;

import android.support.annotation.NonNull;

import com.phicdy.mycuration.data.db.DatabaseAdapter;
import com.phicdy.mycuration.data.rss.Feed;
import com.phicdy.mycuration.presentation.view.RegisterFilterView;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

import java.util.ArrayList;

import static org.junit.Assert.*;

public class RegisterFilterPresenterTest {

    private DatabaseAdapter mockAdapter;
    private RegisterFilterPresenter presenter;
    private MockView mockView;

    @Before
    public void setup() {
        mockAdapter = Mockito.mock(DatabaseAdapter.class);
        presenter = new RegisterFilterPresenter(mockAdapter, 1);
        mockView = new MockView();
        presenter.setView(mockView);
    }

    @Test
    public void setOneSelectedFeedList() {
        ArrayList<Feed> testFeeds = new ArrayList<>();
        testFeeds.add(new Feed(1, "testFeed"));
        presenter.setSelectedFeedList(testFeeds);
        assertEquals(mockView.filterTarget, "testFeed");
    }

    @Test
    public void setMultipleSelectedFeedList() {
        ArrayList<Feed> testFeeds = new ArrayList<>();
        testFeeds.add(new Feed(1, "testFeed"));
        testFeeds.add(new Feed(2, "testFeed2"));
        presenter.setSelectedFeedList(testFeeds);
        assertEquals(mockView.filterTarget, MockView.MULTIPLE_FILTER_TARGET);
    }

    @Test
    public void setEmptySelectedFeedList() {
        ArrayList<Feed> testFeeds = new ArrayList<>();
        presenter.setSelectedFeedList(testFeeds);
        assertEquals(mockView.filterTarget, MockView.DEFAULT_FILTER_TARGET);
    }

    @Test
    public void setNullSelectedFeedList() {
        presenter.setSelectedFeedList(null);
        assertEquals(mockView.filterTarget, MockView.DEFAULT_FILTER_TARGET);
    }

    private class MockView implements RegisterFilterView {
        private String title;
        private String filterTarget = DEFAULT_FILTER_TARGET;
        private static final String DEFAULT_FILTER_TARGET = "default";
        private static final String MULTIPLE_FILTER_TARGET = "multiple-target";

        @Override
        public String filterKeyword() {
            return null;
        }

        @Override
        public String filterUrl() {
            return null;
        }

        @Override
        public String filterTitle() {
            return title;
        }

        @Override
        public void setFilterTitle(@NonNull String title) {
            this.title = title;
        }

        @Override
        public void setFilterTargetRss(@NonNull String rss) {
            this.filterTarget = rss;
        }

        @Override
        public void setMultipleFilterTargetRss() {
            this.filterTarget = MULTIPLE_FILTER_TARGET;
        }

        @Override
        public void resetFilterTargetRss() {
            this.filterTarget = DEFAULT_FILTER_TARGET;
        }

        @Override
        public void setFilterUrl(@NonNull String url) {

        }

        @Override
        public void setFilterKeyword(@NonNull String keyword) {

        }

        @Override
        public void handleEmptyTitle() {

        }

        @Override
        public void handleEmptyCondition() {

        }

        @Override
        public void handlePercentOnly() {

        }

        @Override
        public void finish() {

        }

        @Override
        public void showSaveSuccessToast() {

        }

        @Override
        public void showSaveErrorToast() {

        }

        @Override
        public void trackEdit() {

        }

        @Override
        public void trackRegister() {

        }
    }
}

File: mycuration/src/test/java/com/phicdy/mycuration/presentation/presenter/RegisterFilterPresenterTest.kt
package com.phicdy.mycuration.presentation.presenter

import com.phicdy.mycuration.data.db.DatabaseAdapter
import com.phicdy.mycuration.data.rss.Feed
import com.phicdy.mycuration.presentation.view.RegisterFilterView
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import java.util.ArrayList

class RegisterFilterPresenterTest {

    private lateinit var mockAdapter: DatabaseAdapter
    private lateinit var presenter: RegisterFilterPresenter
    private lateinit var mockView: MockView

    @Before
    fun setup() {
        mockAdapter = Mockito.mock(DatabaseAdapter::class.java)
        mockView = MockView()
        presenter = RegisterFilterPresenter(mockView, mockAdapter, 1)
    }

    @Test
    fun `when set one selected feed list then filter target is the one`() {
        val testFeeds = ArrayList<Feed>()
        testFeeds.add(Feed(1, "testFeed", "", Feed.DEDAULT_ICON_PATH, "", 0, ""))
        presenter.setSelectedFeedList(testFeeds)
        assertEquals(mockView.filterTarget, "testFeed")
    }

    @Test
    fun `when set two selected feed list then filter target is multiple target`() {
        val testFeeds = ArrayList<Feed>()
        testFeeds.add(Feed(1, "testFeed", "", Feed.DEDAULT_ICON_PATH, "", 0, ""))
        testFeeds.add(Feed(2, "testFeed2", "", Feed.DEDAULT_ICON_PATH, "", 0, ""))
        presenter.setSelectedFeedList(testFeeds)
        assertEquals(mockView.filterTarget, MockView.MULTIPLE_FILTER_TARGET)
    }

    @Test
    fun `when set empty selected feed list then filter target is default`() {
        val testFeeds = ArrayList<Feed>()
        presenter.setSelectedFeedList(testFeeds)
        assertEquals(mockView.filterTarget, MockView.DEFAULT_FILTER_TARGET)
    }

    private class MockView : RegisterFilterView {
        private var title: String? = null
        var filterTarget = DEFAULT_FILTER_TARGET

        override fun filterKeyword(): String {
            return ""
        }

        override fun filterUrl(): String {
            return ""
        }

        override fun filterTitle(): String {
            return ""
        }

        override fun setFilterTitle(title: String) {
            this.title = title
        }

        override fun setFilterTargetRss(rss: String) {
            this.filterTarget = rss
        }

        override fun setMultipleFilterTargetRss() {
            this.filterTarget = MULTIPLE_FILTER_TARGET
        }

        override fun resetFilterTargetRss() {
            this.filterTarget = DEFAULT_FILTER_TARGET
        }

        override fun setFilterUrl(url: String) {

        }

        override fun setFilterKeyword(keyword: String) {

        }

        override fun handleEmptyTitle() {

        }

        override fun handleEmptyCondition() {

        }

        override fun handlePercentOnly() {

        }

        override fun finish() {

        }

        override fun showSaveSuccessToast() {

        }

        override fun showSaveErrorToast() {

        }

        override fun trackEdit() {

        }

        override fun trackRegister() {

        }

        companion object {
            const val DEFAULT_FILTER_TARGET = "default"
            const val MULTIPLE_FILTER_TARGET = "multiple-target"
        }
    }
}